### PR TITLE
Revised config and code quality improvements

### LINKED ISSUE #1534: Convert model config from flat to hierarchical dict
### Describe the task. It can be a feature, documentation, etc.

Convert config from flat to hierarchical dict

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

_No response_

### Area

- [ ] datasets, data readers, data preparation and transfer
- [x] model
- [ ] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation

### LINKED ISSUE #1535: Batch_size (num_samples) > 1 is currently broken
### What happened?

Batch_size (num_samples) > 1 is currently broken

### What are the steps to reproduce the bug?

Train with num_samples > 1. Loss is not converging.

### Hedgedoc link to logs and more information. This ticket is public, do not attach files directly.

_No response_

### PR DESCRIPTION
**Pull request by clessig**
## Description

Revise config to nested dict; simplify code where possible and where changes are necessary anyway.

This PR also enables a more flexible combination of different loss terms, e.g. of a physical space and latent loss, as demonstrated in the default config. It also decouples training and validation and test as much as possible, so that one can have different objectives for these.

## Issue Number

Closes #1534 
Closes #1535

## Checklist before asking for review

-   [ ] I have performed a self-review of my code
-   [ ] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [ ] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

**Comment by clessig** [comment]
forecast, ERA5 + NPP-ATMS + SYNOP


num_samples/batch_size = 1: (520709 : g479xfk6)

0: LossPhysical.ERA5.mse.avg : 9.4735E-02       
0: LossPhysical.ERA5.mae.avg : 1.9303E-01       
0: LossPhysical.NPPATMS.mse.avg : 1.1044E-02    
0: LossPhysical.NPPATMS.mae.avg : 2.9273E-02    
0: LossPhysical.SurfaceCombined.mse.avg : 2.6745E-01    
0: LossPhysical.SurfaceCombined.mae.avg : 2.8849E-01    
0: LossPhysical.loss_avg : 1.8611E-01 



num_samples/batch_size = 2: 520711 : ypix0nr7

0: LossPhysical.ERA5.mse.avg : 1.1328E-01       
0: LossPhysical.ERA5.mae.avg : 2.1473E-01       
0: LossPhysical.NPPATMS.mse.avg : 1.8830E-02    
0: LossPhysical.NPPATMS.mae.avg : 4.5836E-02    
0: LossPhysical.SurfaceCombined.mse.avg : 2.9414E-01    
0: LossPhysical.SurfaceCombined.mae.avg : 3.0358E-01    
0: LossPhysical.loss_avg : 2.0496E-01

File: config/default_config.yml [hunk 1]
```diff
@@ -45,24 +44,19 @@ pred_adapter_kv: False
 pred_self_attention: True
 pred_dyadic_dims: False
 pred_mlp_adaln: True
-num_class_tokens: 1
-num_register_tokens: 7
+num_class_tokens: 0
```
Comment (sophie-xhonneux) [inline config/default_config.yml]: I think the default register tokens should probably not be 0

File: config/default_config.yml [hunk 2]
```diff
@@ -45,24 +44,19 @@ pred_adapter_kv: False
...
-forecast_freeze_model: False
-forecast_att_dense_rate: 1.0
-forecast_with_step_conditioning: True # False
 fe_num_blocks: 0
 fe_num_heads: 16
 fe_dropout_rate: 0.1
 fe_with_qk_lnorm: True
 fe_layer_norm_after_blocks: []  # Index starts at 0. Thus, [3] adds a LayerNorm after the fourth layer
 fe_impute_latent_noise_std: 0.0  # 1e-4
+forecast_att_dense_rate: 1.0
```
Comment (sophie-xhonneux) [inline config/default_config.yml:65]: I think maybe adding a comment here or in the code would be good to explain that this can cause an error when not 1.0 due to local attention not existing

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 3]
```diff
@@ -71,45 +71,54 @@ class MultiStreamDataSampler(torch.utils.data.IterableDataset):
     def __init__(
         self,
         cf: Config,
-        start_date: np.datetime64,
-        end_date: np.datetime64,
-        batch_size: int,
-        samples_per_mini_epoch: int,
+        mode_cfg: dict,
```
Comment (sophie-xhonneux) [inline src/weathergen/datasets/multi_stream_data_sampler.py:74]: I personally find this name unintuitive. Maybe we can at least a comment explain why we need cf and mode_cfg and what each contains?

File: src/weathergen/model/engines.py [hunk 4]
```diff
@@ -91,32 +92,40 @@ def forward(self, batch, pe_embed):
...
-            scatter_idxs = scatter_idxs.unsqueeze(1).repeat((1, self.cf.ae_local_dim_embed))
-            pe_idxs = torch.cat(pe_idxs)
-
             # embedding from physical space to per patch latent representation
-            x_embed = self.embeds[stream_name](sdata).flatten(0, 1)
+            x_embeds += [self.embeds[stream_name](sdata).flatten(0, 1)]
+
+        # switch from stream to cell-based ordering and apply per cell positional encoding
+
+        # computer scatter index across batch items and input steps
```
Comment (sophie-xhonneux) [inline src/weathergen/model/engines.py:113]: I might be out of the loop, how come we changed the ordering from stream to cell based ordering?

File: src/weathergen/model/engines.py [hunk 5]
```diff
@@ -395,7 +404,7 @@ def __init__(self, cf: Config, num_healpix_cells: int, dim_aux: int = None) -> N
         self.fe_blocks = torch.nn.ModuleList()
 
         global_rate = int(1 / self.cf.forecast_att_dense_rate)
-        if self.cf.forecast_policy is not None:
+        if mode_cfg.get("forecast", {}).get("policy") is not None:
```
Comment (sophie-xhonneux) [inline src/weathergen/model/engines.py:407]: should we have a default for `.get("policy")`

File: src/weathergen/model/model_interface.py [hunk 6]
```diff
@@ -256,20 +257,20 @@ def get_model(cf: Config, training_mode: TrainingMode, dataset, overrides):
     ).create()
 
 
-def get_target_aux_calculator(cf: Config, dataset, model, device, **kwargs):
+def get_target_aux_calculator(
+    cf: Config, cfg: OmegaConf, dataset, model, device, batch_size_per_gpu, **kwargs
```
Comment (sophie-xhonneux) [inline src/weathergen/model/model_interface.py]: the `cf` and `cfg` is confusing, we should think if we have a better alternative

File: src/weathergen/model/model_interface.py [hunk 7]
```diff
@@ -279,9 +280,9 @@ def get_target_aux_calculator(cf: Config, dataset, model, device, **kwargs):
             is_model_sharded=(cf.with_ddp and cf.with_fsdp),
         )
 
-        target_aux = EMATeacher(
-            model, ema_model, get_batch_size(cf, cf.world_size_original), **cf.training_config
-        )
+        batch_size = cf.get("world_size_original", cf.get("world_size")) * batch_size_per_gpu
```
Comment (sophie-xhonneux) [inline src/weathergen/model/model_interface.py:303]: shouldn't we use `get_batch_size_from_config` here like in the trainer?

**Comment by clessig** [comment]
Minimal config in `training_config`:

```
  losses : {
    "physical": {
        type: LossPhysical,
        loss_fcts: { "mse": { }, },
        },
    }
  
  model_input: {
    "forecasting" : {
      masking_strategy: "forecast",
    }
  }
```

File: src/weathergen/model/engines.py (hunk 5 above)
Comment (clessig) [reply src/weathergen/model/engines.py:407]: Default is `null` in yaml, i.e. is None. In this case forecasting is disabled. I think that's sensible.

File: config/default_config.yml (hunk 1 above)
Comment (clessig) [reply config/default_config.yml]: I think the defaul_config should be a relatively minimal example to run the model. So I don't think we need register tokens. But we can have a discussion in #1561

File: src/weathergen/datasets/multi_stream_data_sampler.py (hunk 3 above)
Comment (clessig) [reply src/weathergen/datasets/multi_stream_data_sampler.py:74]: I agree but I couldn't think of a term that captures training/test/validation

File: src/weathergen/model/engines.py (hunk 4 above)
Comment (clessig) [reply src/weathergen/model/engines.py:113]: This always happened here. Embedding is per stream since it happens independent for each cell and stream, subsequent computations combine tokens from different streams per cell.

File: src/weathergen/model/model_interface.py (hunk 6 above)
Comment (clessig) [reply src/weathergen/model/model_interface.py]: Yes, I changed the name to `loss_term_cfg`.

File: src/weathergen/model/model_interface.py (hunk 7 above)
Comment (clessig) [reply src/weathergen/model/model_interface.py:303]: No, because we can have a different batch size for training and validation. We pass the correct one as an argument.

**Review by sophie-xhonneux — APPROVED** [review]

**Comment by clessig** [comment]
Default integration test is working. train_continue and inference are also working.
